File: EdDyA II/SinglyLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
    def __str__(self):
        s = str(self.data)
        s += "\t&: " + str(hex(id(self)))
        s += "\tNext: " + str(hex(id(self.next)))
        return s
class SinglyLinkedList:
    def __init__(self):
        self.first = None
    def insert_head(self, data):
        n = Node(data)
        n.next = self.first
        self.first = n
    def insert_tail(self, data):
        n = Node(data)
        if self.first is None:
            self.first = n
        else:
            aux = self.first
            while aux.next is not None:
                aux = aux.next
            aux.next = n
    def remove(self, data):
        if self.first is None: return
        aux = Node(None)
        aux.next = self.first
        while aux.next is not None and aux.next.data is not data:
            aux = aux.next
        if aux.next is self.first: self.first = self.first.next
        if aux.next is not None: aux.next = aux.next.next
        else: return -1
    def __str__(self):
        if self.first is None: return "[]"
        s, aux = "", self.first
        while aux is not None:
            s += str(aux) + '\n'
            aux = aux.next
        return s
l = SinglyLinkedList()
l.insert_head(5)
l.insert_tail(10)
l.remove(5)
l.insert_head(5)
l.remove(10)
l.insert_tail(10)
l.remove(7)
print(l)

File: EdDyA II/Lists/SinglyLinkedList.cpp
//Enteros
#include <iostream>
using namespace std;
class Node{
  public:
    int data;
    Node *next;
    Node(int data){
      cout << "\tNode(" << data << ")" << endl;
      this -> data = data;
      this -> next = NULL; //nullptr
    }
    ~Node(){
      cout << "\t~Node(" << data << ")" << endl;
    }
    friend ostream& operator << (ostream& output, const Node &n){
      output << "Data: " << n.data << "\tNext: " << n.next << endl;
    }
};
class SinglyLinkedList{
  public:
    Node* head;
    SinglyLinkedList(){
      head = NULL;
      cout << "\tSinglyLinkedList()" << endl;
    }
    ~SinglyLinkedList(){
      cout << "\t~SinglyLinkedList()" << endl;
      Node *a = head;
      while( a != NULL ){
	Node *b = a -> next;
	delete a;
	a = b;
      }
    }
    void insert_tail(int data){
      Node *n = new Node(data);
      if( head == NULL ) head = n;
      else{
        Node *a = head;
        while( a -> next != NULL ) a = a -> next;
        a -> next = n;
      }
    }
    friend ostream &operator << (ostream &output, const SinglyLinkedList &l){
      Node *n = l.head;
      while( n != NULL ){
	output << *n;
	n = n -> next;
      }
//      output << endl;
    }
};
int main(){
  SinglyLinkedList lista;
  lista.insert_tail(1);
  lista.insert_tail(2);
  lista.insert_tail(3);
  cout << lista;
  return 0;
}
